### PR TITLE
Add Geotagged Image Support

### PR DESCRIPTION
**Pull request by jbuckner**
This adds support for dragging geotagged jpg images and puts markers on the map. When you click on the marker, you get a popup with the image.

It uses [`exif-js`](https://github.com/exif-js/exif-js) for EXIF parsing.

derive_image2

### CODE REVIEW

**Comment by erik** [comment]
Awesome, thanks for the PR!

I won't have time to review until after the weekend but the idea looks cool.

File: src/ui.js [hunk 1]
```diff
@@ -239,6 +272,29 @@ export function buildSettingsModal(tracks, opts, finishCallback) {
             value=${opts.lineOptions.weight}>
     </span>
 
+    <span class="form-row">
```
Comment (erik) [inline src/ui.js]: Maybe these should be grouped under "image settings" or so? Whatever you think makes it more clear

File: src/ui.js [hunk 2]
```diff
@@ -77,29 +77,62 @@ function handleFileSelect(map, evt) {
...
+            parseGPX(reader.result)
+                .then(parsedTracks => {
+                    parsedTracks.forEach(track => {
+                        track.filename = file.name;
+                        tracks.push(track);
+                        map.addTrack(track);
+                        modal.addSuccess();
+                    });
+                })
+                .then(resolve);
```
Comment (xsanda) [inline src/ui.js]: Again, wrapped promise not needed: .then() returns a promise that resolves when it completes/when the promise its code returns resolves

File: src/ui.js [hunk 3]
```diff
@@ -77,29 +77,62 @@ function handleFileSelect(map, evt) {
...
-        })
-        .catch(err => {
-            modal.addFailure({name: file.name, error: err});
+    let handleImage = file => new Promise(resolve => {
+        let image = new Image(file);
+
+        image.extractExifData().then((image) => {
+            if (!image) {
+                modal.addFailure({name: file.name, error: 'No geolocation data'});
+                return resolve();
```
[the remark this answers is not in the document]
Comment (xsanda) [reply src/ui.js]: This whole function would be nicer as an async:

```js
let handleImage = async (file) => {
    let image = await new Image(file).extractExitData();
    if (!image) throw 'No geolocation data');
    map.addImage(image);
    modal.addSuccess();
}

File: src/ui.js [hunk 4]
```diff
@@ -77,29 +77,62 @@ function handleFileSelect(map, evt) {
...
+                break;
+            case 'jpg':
+                handleImage(file)
+                    .then(resolve);
+                break;
+            default:
+                console.log(`File ${file.name} is an unsupported format.`);
+                modal.addFailure({name: file.name, error: 'Unsupported file format'});
+                resolve();
+        }
```
Comment (xsanda) [inline src/ui.js]:
```javascript
let detectFileType = async file => {
    let extension = ...
    try {
        switch (extension) {
            case 'gpx':
                return await handleGpx(file);
            case 'jpg':
            case 'jpeg':
                return await handleImage(file);
            default:
                throw 'Unsupported file format';
        }
    } catch (err) {
        modal.addFailure(…);
    }
}
```
Shouldn't this be called handleFile, as it does more than just detecting what the file type is, it also handles it based on said file type

File: src/map.js [hunk 5]
```diff
@@ -163,6 +183,36 @@ export default class GpxMap {
         this.tracks.push(line);
     }
 
+    markerClick(image) {
+        let latlng = leaflet.latLng(image.latitude, image.longitude);
+
+        return new Promise(resolve => {
+            image.getImageData().then((imageData) => {
+                leaflet.popup(latlng)
+                    .setLatLng(latlng)
```
Comment (xsanda) [inline src/map.js]: Why does it need to receive latlng twice?

File: src/ui.js [hunk 6]
```diff
@@ -77,29 +77,62 @@ function handleFileSelect(map, evt) {
 
     modal.show();
 
-    let openFile = file => new Promise(resolve => {
+    let handleGpx = file => new Promise(resolve => {
```
Comment (xsanda) [inline src/ui.js]: Can this be named `handleGPX` to match `parseGPX`?

File: src/ui.js [hunk 7]
```diff
@@ -77,29 +77,62 @@ function handleFileSelect(map, evt) {
...
+                .then(parsedTracks => {
+                    parsedTracks.forEach(track => {
+                        track.filename = file.name;
+                        tracks.push(track);
+                        map.addTrack(track);
+                        modal.addSuccess();
+                    });
+                })
+                .then(resolve);
+        };
```
Comment (xsanda) [inline src/ui.js]:
```javascript
const parseGPX = async file => {
    const fileContents = await new Promise(resolve => {
        const reader = new FileReader();
        reader.onload = () => resolve(reader.result);
        reader.readAsText(file, 'UTF-8');
    });
    for (const track of await parseGPX(fileContents)) {
        track.filename = file.name;
        tracks.push(track);
        map.addTrack(track);
        modal.addSuccess();
    }
});
```

File: src/image.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    return dms[0].numerator + dms[1].numerator /
+        (60 * dms[1].denominator) + dms[2].numerator / 
+        (3600 * dms[2].denominator);
+}
+
+export default class Image {
+    constructor(imageFile) {
+        this.imageFile = imageFile;
+        this.imageData = null;
+        this.hasParsedExif = false;
```
[the remark this answers is not in the document]
Comment (jbuckner) [reply src/image.js]: I should've named `imageData` better. `imageData` is a cache of the base64 encoded image representation that we get from `reader.readAsDataURL(this.imageFile, 'UTF-8');` in `getImageData()` below.

I was a bit torn on caching the image data at all or just reading from the file every time. The memory usage and performance looked pretty similar with either approach. I can just remove the `imageData` cache altogether.

**Comment by jbuckner** [comment]
I think I got to all of the feedback. I had to add `babel-polyfill` to support `async` / `await` syntax. Man those are nice!

Unfortunately it looks like `jshint` doesn't support `async` / `await` so now Travis is failing. I'm seeing what options are available to fix this. It looks like there's a linter called [`eslint`](https://eslint.org) that supports it but I need to dig into this a bit more.

**Comment by erik** [comment]
Nice! Thanks for addressing our earlier comments.

This looks good to me, I can figure out switching over to eslint as a follow up PR, I've used eslint before and I'm fine migrating this project over to it.

**Comment by jbuckner** [comment]
Great, thanks! Let me merge `master` in and fix linting in my branch and I'll push.

**Comment by erik** [comment]
Oops, sorry for stepping on your toes there with the merge commit. But yeah, the errors at this point are style only, so we should be good to go after that 👍

**Comment by jbuckner** [comment]
No worries. I just fixed the last few lint issues so the next Travis run should be good.

**Pull request merged**